source/Body.cpp

```diff
@@ -388,60 +819,150 @@ void Body::SetStep(int step) const
 	// and caches the mask and the frame so that if further queries are made at
 	// this same time step, we don't need to redo the calculations.
 	if(step == currentStep || step < 0 || !sprite || !sprite->Frames())
-		return;
-	currentStep = step;
+		return false;
 
 	// If the sprite only has one frame, no need to animate anything.
 	float frames = sprite->Frames();
 	if(frames <= 1.f)
 	{
 		frame = 0.f;
-		return;
+		return false;
 	}
 	float lastFrame = frames - 1.f;
 	// This is the number of frames per full cycle. If rewinding, a full cycle
 	// includes the first and last frames once and every other frame twice.
-	float cycle = (rewind ? 2.f * lastFrame : frames) + delay;
+	float cycle = (anim.rewind ? 2.f * lastFrame : frames) + anim.delay;
 
 	// If this is the very first step, fill in some values that we could not set
 	// until we knew the sprite's frame count and the starting step.
-	if(randomize)
+	if(anim.randomizeStart)
 	{
-		randomize = false;
+		anim.randomizeStart = false;
 		// The random offset can be a fractional frame.
-		frameOffset += static_cast<float>(Random::Real()) * cycle;
+		integratedFrame = static_cast<float>(Random::Real()) * cycle;
 	}
-	else if(startAtZero)
+	else if(anim.startAtZero)
 	{
-		startAtZero = false;
-		// Adjust frameOffset so that this step's frame is exactly 0 (no fade).
-		frameOffset -= frameRate * step;
+		anim.startAtZero = false;
+		// Adjust integrated frame to start at the preferred start frame.
+		integratedFrame = anim.startFrame;
 	}
 
+	// Clamp frameRate from previous calculations.
+	if(frameRate < Body::MIN_FRAME_RATE)
+		frameRate = Body::MIN_FRAME_RATE;
+
+	if(frameRate > anim.frameRate)
+		frameRate = anim.frameRate;
+
 	// Figure out what fraction of the way in between frames we are. Avoid any
 	// possible floating-point glitches that might result in a negative frame.
-	frame = max(0.f, frameRate * step + frameOffset);
-	// If repeating, wrap the frame index by the total cycle time.
-	if(repeat)
-		frame = fmod(frame, cycle);
-
-	if(!rewind)
+	int prevFrame = static_cast<int>(frame), nextFrame = -1;
```

Comment:
I'm not sure if this is part of the official style guide, but we usually put every variable declaration on a separate line to improve code readability. Also I think that the explicit cast isn't necessary in this case.
```suggestion
	int prevFrame = frame;
	int nextFrame = -1;
```